### PR TITLE
feat(api-v2): Add an RDF processing façade (DSP-1020)

### PR DESCRIPTION
**Pull request by benjamingeer**
This PR adds an RDF processing façade in `webapi/src/main/scala/org/knora/webapi/messages/util/rdf`, with two different implementations (Jena and RDF4J).

## The API

- `RdfModel`, which represents a set of RDF graphs (a default graph and/or one or more named graphs)
- `RdfNode` and its subclasses, which represent RDF nodes (IRIs, blank nodes, and literals)
- `Statement`, which represents a triple or quad
- `RdfNodeFactory`, which creates nodes and statements
- `RdfModelFactory`, which creates empty RDF models
- `RdfFormatUtil`, which parses and formats RDF
- `RdfFeatureFactory`, which returns instances of `RdfNodeFactory`, `RdfModelFactory`, and `RdfFormatUtil`, using feature toggle configuration.

## The implementations

- The Jena-based implementation, in package `org.knora.webapi.messages.util.rdf.jenaimpl`
- The RDF4J-based implementation, in package `org.knora.webapi.messages.util.rdf.rdf4jimpl`

## The feature toggle

`jena-rdf-library`:

- `on` means use the Jena implementation
- `off` (the default) means use the RDF4J implementation, which was previously the main one used in Knora

## Tasks

- [x] Add traits for the façade.
- [x] Wrap model building and querying functionality (`RdfModel`, `RdfModelFactory`, `RdfNodeFactory`).
- [x] Add a feature toggle and feature factory (`RdfFeatureFactory`).
- [x] Wrap formatting and parsing (`RdfFormatUtil`).
- [x] Use the façade:
  - [x] `JsonLDUtil`
  - [x] `RouteUtilV2`
  - [x] `KnoraRequestV2`
  - [x] `KnoraResponseV2`
  - [x] `HttpTriplestoreConnector`:
    - [x] `SparqlConstructRequest`
    - [x] `SparqlExtendedConstructRequest`
- [x] Add `featureFactoryConfig` to everything that depends on `CONSTRUCT` requests (most of the changes in this PR)
- [x] Add abstract test classes, with subclasses that test using Jena and RDF4J:
  - [x] `RDF4JModelSpec`
  - [x] `RDF4JFormatUtilSpec`
  - [x] `JsonLDUtilSpec`
  - [x] `KnoraResponseV2Spec`
- [x] Update tests:
  - [x] `E2ESpec` and subclasses.
  - [x] `R2RSpec` and subclasses.
  - [x] `MetadataMessagesV2Spec`
  - [x] `MetadataRouteV2E2ESpec`
- [x] Add docs.
- [x] Clean up Bazel dependencies.

## What still uses RDF4J directly

- Things that use RDF4J's streaming API to process large amounts of data, especially to avoid constructing a large string in TriG format:
  - `ProjectsResponderADM.projectDataGetRequestADM`
  - `HttpTriplestoreConnector.turtleToTrig`
  - `RepositoryUpdater`
- The repository update plugin tests, which use SPARQL
- `TEIHeader`: uses XSLT that depends on the exact format of RDF/XML generated by RDF4J. The XSLT would need to be improved to handle `rdf:Description`.
- `GravsearchParser`: uses RDF4J's SPARQL parser, not worth changing

## TODO in a later PR

- SHACL validation
- SPARQL querying
- A streaming parsing/formatting API for processing very large graphs

### CODE REVIEW

File: webapi/src/main/scala/org/knora/webapi/messages/util/rdf/jenaimpl/JenaModel.scala [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+    override def isIsomorphicWith(otherRdfModel: RdfModel): Boolean = {
+        // Jena's DatasetGraph doesn't have a method for this, so we have to do it ourselves.
+
+        val thatDatasetGraph: jena.sparql.core.DatasetGraph = otherRdfModel.asJenaDataset.asDatasetGraph
+
+        // Get the IRIs of the named graphs.
+        val thisModelNamedGraphIris: Set[jena.graph.Node] = datasetGraph.listGraphNodes.asScala.toSet
+        val thatModelNamedGraphIris: Set[jena.graph.Node] = thatDatasetGraph.listGraphNodes.asScala.toSet
+
+        // The two models are isomorphic if:
```
Comment (SepidehAlassi) [inline webapi/src/main/scala/org/knora/webapi/messages/util/rdf/jenaimpl/JenaModel.scala:292]: Nice! without this explanation, I wouldn't have got this part.

File: webapi/src/test/scala/org/knora/webapi/messages/util/rdf/RdfModelSpec.scala [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+        "add and find quads" in {
+            val labelPred: IriNode = nodeFactory.makeIriNode(OntologyConstants.Rdfs.Label)
+            val commentPred: IriNode = nodeFactory.makeIriNode(OntologyConstants.Rdfs.Comment)
+            val context1 = "http://example.org/graph1"
+
+            val graph1LabelStatement = nodeFactory.makeStatement(
+                subj = nodeFactory.makeIriNode("http://example.org/6"),
+                pred = labelPred,
+                obj = nodeFactory.makeDatatypeLiteral(value = "Lucky's Discount X-Wing Repair", datatype = OntologyConstants.Xsd.String),
```
Comment (SepidehAlassi) [inline webapi/src/test/scala/org/knora/webapi/messages/util/rdf/RdfModelSpec.scala:171]: :-D

File: webapi/src/test/scala/org/knora/webapi/messages/util/rdf/RdfModelSpec.scala [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+                graph1LabelStatement,
+                graph1CommentStatement
+            )
+
+            val context2 = "http://example.org/graph2"
+
+            val graph2LabelStatement = nodeFactory.makeStatement(
+                subj = nodeFactory.makeIriNode("http://example.org/7"),
+                pred = labelPred,
+                obj = nodeFactory.makeDatatypeLiteral(value = "Mos Eisley Used Droids", datatype = OntologyConstants.Xsd.String),
```
Comment (SepidehAlassi) [inline webapi/src/test/scala/org/knora/webapi/messages/util/rdf/RdfModelSpec.scala:192]: "Mos Eisley used droids that were not allowed in the Cantina"

File: webapi/src/test/scala/org/knora/webapi/messages/util/rdf/JsonLDUtilSpec.scala [hunk 4]
```diff
@@ -223,10 +224,60 @@ abstract class JsonLDUtilSpec(featureToggle: FeatureToggle) extends CoreSpec {
...
 
             // Parse the same file to an RDF4J Model.
-            val jsonLDExpectedModel: RdfModel = rdfFormatUtil.parseToRdfModel(expectedJsonLD, JsonLD)
+            val jsonLDExpectedModel: RdfModel = rdfFormatUtil.parseToRdfModel(rdfStr = expectedJsonLD, rdfFormat = JsonLD)
 
             // Compare that with the model generated by the JsonLDDocument.
             jsonLDOutputModel should ===(jsonLDExpectedModel)
         }
+
+        "correctly convert an RDF model to JSON-LD if it contains a circular reference" in {
```
Comment (SepidehAlassi) [inline webapi/src/test/scala/org/knora/webapi/messages/util/rdf/JsonLDUtilSpec.scala:232]: Excellent!

**Review by SepidehAlassi — APPROVED** [review]

**Comment by SepidehAlassi** [comment]
@benjamingeer This looks great, thanks for all your work.

**Comment by benjamingeer** [comment]
Many thanks for reviewing this!

**Pull request merged**